x/ratelimit/keeper/packet.go

```diff
@@ -102,65 +112,120 @@ func ParseDenomFromRecvPacket(packet channeltypes.Packet, packetData transfertyp
+}
+
+// Middleware implementation for SendPacket with rate limiting
+// Checks whether the rate limit has been exceeded - and if it hasn't, sends the packet
+func (k Keeper) SendRateLimitedPacket(ctx sdk.Context, packet channeltypes.Packet) error {
+	packetInfo, err := ParsePacketInfo(packet, types.PACKET_SEND)
+	if err != nil {
+		return err
+	}
 
-	err := k.CheckRateLimitAndUpdateFlow(ctx, types.PACKET_SEND, denom, channelId, amount)
+	// Check if the packet would exceed the outflow rate limit
+	updatedFlow, err := k.CheckRateLimitAndUpdateFlow(ctx, types.PACKET_SEND, packetInfo)
 	if err != nil {
 		return err
 	}
 
+	// Store the sequence number of the packet so that if the transfer fails,
+	// we can identify if it was sent during this quota and can revert the outflow
+	if updatedFlow {
+		k.SetPendingSendPacket(ctx, packetInfo.ChannelID, packet.Sequence)
+	}
+
 	return nil
 }
 
 // Middleware implementation for RecvPacket with rate limiting
+// Checks whether the rate limit has been exceeded - and if it hasn't, allows the packet
 func (k Keeper) ReceiveRateLimitedPacket(ctx sdk.Context, packet channeltypes.Packet) error {
-	// The Stride channelID should always be used as the key for the RateLimit object (not the counterparty channelID)
-	// For a RECEIVE packet, the Stride channelID is the DESTINATION channel
-	// This is because the Source and Desination are defined from the perspective of a packet recipient
-	// Meaning, when this packet lands on a Stride, the "Source" will be the host zone's channel,
-	//  and the "Destination" will be the Stride Channel
-	channelId := packet.GetDestChannel()
-
-	// Parse the amount and denom from the packet
-	var packetData transfertypes.FungibleTokenPacketData
-	if err := json.Unmarshal(packet.GetData(), &packetData); err != nil {
+	packetInfo, err := ParsePacketInfo(packet, types.PACKET_RECV)
+	if err != nil {
 		return err
 	}
 
-	amount, ok := sdk.NewIntFromString(packetData.Amount)
-	if !ok {
-		return errorsmod.Wrapf(sdkerrors.ErrInvalidRequest, "Unable to cast packet amount to sdkmath.Int")
+	_, err = k.CheckRateLimitAndUpdateFlow(ctx, types.PACKET_RECV, packetInfo)
+	return err
+}
+
+// Middleware implementation for OnAckPacket with rate limiting
+// If the packet failed, we should decrement the Outflow
+func (k Keeper) AcknowledgeRateLimitedPacket(ctx sdk.Context, packet channeltypes.Packet, acknowledgement []byte) error {
+	// Check whether the ack was a success or error
+	isICA := false
+	ackResponse, err := icacallbacks.UnpackAcknowledgementResponse(ctx, k.Logger(ctx), acknowledgement, isICA)
+	if err != nil {
+		return err
 	}
 
-	denom := ParseDenomFromRecvPacket(packet, packetData)
+	// Parse the denom, channelId, and amount from the packet
+	packetInfo, err := ParsePacketInfo(packet, types.PACKET_SEND)
+	if err != nil {
+		return err
+	}
+
+	// If the ack was successful, remove the pending packet
+	if ackResponse.Status == icacallbacktypes.AckResponseStatus_SUCCESS {
```

Comment:
Should this check if the packet exists before removing it? e.g. what if it's just a normal ack (from a channel that has no rate limit)

Reply:
the stored pending packets is only used for rate limiting to determine if the packet was sent in the same rate limit window. This code isn't deleting the actual packet, just the record that it was sent this window. It has no affect on other modules

Reply:
That makes sense, but I think we only set the pending packet if we update a flow
```
	if updatedFlow {
		k.SetPendingSendPacket(ctx, packetInfo.ChannelID, packet.Sequence)
	}
```

However, this will always try and delete a packet (unless I'm missing something). If you call `Delete` on a key that doesn't exist, does it throw an error?

Reply:
Oh right, sorry I misunderstood your original question. Yeah this was intentional since calling Delete on a key that doesn't exist does not error 

I figured it was simpler to just call Remove regardless, but if you think it would help readability, I can add the check that it exists first before removing

Reply:
Gotcha! Makes sense, either way seems fine to me

Reply:
cool, I'll just leave it as is